### PR TITLE
Deep scan code changes

### PR DESCRIPTION
**Pull request by v-himkam**
Checks added : 

1. find inactive extensions
2. display extensions with critical scopes
3. Avoid extensions  with 'DevTest', 'Demo', 'Preview', 'Deprecated' in names
4.  Display extensions with Top Publishers 
5. Display extensions that are private 
6. Display Nonprod extensions

### CODE REVIEW

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 1]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
 
                 if($extCount -gt 0)
-                {               
-                    $controlResult.AddMessage("No. of installed extensions: " + $extCount);
+                {   
+                    if([AzSKRoot]::IsDetailedScanRequired -eq $true)
+                    {
+                        # find inactive extensions
+                        $date = Get-Date
+                        $ExtensionsLastUpdatedInYears=$this.ControlSettings.Organization.ExtensionsLastUpdatedInYears;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:416]: add check for control settings...

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 2]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
                 $extensionList = $extensionList | Where-Object {$_.flags -notlike "*builtin*" } # to filter out extensions that are built in and are not visible on portal.
+                
+
                 $extCount = ($extensionList | Measure-Object ).Count;
 
                 if($extCount -gt 0)
-                {               
-                    $controlResult.AddMessage("No. of installed extensions: " + $extCount);
+                {   
+                    if([AzSKRoot]::IsDetailedScanRequired -eq $true)
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:412]: move it after basic detailed logs is filled

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 3]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                    {
+                        # find inactive extensions
+                        $date = Get-Date
+                        $ExtensionsLastUpdatedInYears=$this.ControlSettings.Organization.ExtensionsLastUpdatedInYears;
+                        $thresholddate = $date.AddYears(-$ExtensionsLastUpdatedInYears)
+                        $staleExtensionList = $extensionList | Where-Object {([datetime] $_.lastPublished) -lt $thresholddate}
+                        $controlResult.AddMessage("`nNo. of extensions not have been updated since past $ExtensionsLastUpdatedInYears years : "+ $staleExtensionList.count)
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:419]: No. of extensions that haven't been published in last $variable years:

Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:419]: add if condition for count

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 4]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                        $thresholddate = $date.AddYears(-$ExtensionsLastUpdatedInYears)
+                        $staleExtensionList = $extensionList | Where-Object {([datetime] $_.lastPublished) -lt $thresholddate}
+                        $controlResult.AddMessage("`nNo. of extensions not have been updated since past $ExtensionsLastUpdatedInYears years : "+ $staleExtensionList.count)                        
+                        $display= $staleExtensionList|Format-Table -Property  @{name="ExtensionName";expression={$_.extensionName}},@{name="PublisherName";expression={$_.publisherName}} | Out-String
+                        $controlResult.AddMessage($display) 
+
+                        # display extensions with critical scopes
+                        $ExtensionCriticalScopes=$this.ControlSettings.Organization.ExtensionCriticalScopes;
+                        $ExtensionListWithCriticalScopes = $extensionList | Where-Object {$ExtensionCriticalScopes -contains $_.scopes }
+                        $controlResult.AddMessage("`nNo. of extensions have critical access permissions : "+ $ExtensionListWithCriticalScopes.count)
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:426]: No. of extensions that have critical access permissions:

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 5]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                        # Avoid extensions  with 'DevTest', 'Demo', 'Preview', 'Deprecated' in names
+                        $ExtensionListWithInvalidExtensionNames=@()
+                        $InvalidExtensionNames=$this.ControlSettings.Organization.InvalidExtensionNames;
+                        for($i=0;$i -lt $extensionList.count;$i++)
+                        {
+                            for($j=0;$j -lt $InvalidExtensionNames.Count;$j++)
+                            {
+                                if($extensionList[$i] -match $InvalidExtensionNames[$j])
+                                {
+                                    [System.Array] $ExtensionListWithInvalidExtensionNames = $extensionList[$i]
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:439]: $ExtensionListWithInvalidExtensionNames += $extensionList[$i]

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 6]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                        $controlResult.AddMessage("`nNo. of extensions not have been updated since past $ExtensionsLastUpdatedInYears years : "+ $staleExtensionList.count)                        
+                        $display= $staleExtensionList|Format-Table -Property  @{name="ExtensionName";expression={$_.extensionName}},@{name="PublisherName";expression={$_.publisherName}} | Out-String
+                        $controlResult.AddMessage($display) 
+
+                        # display extensions with critical scopes
+                        $ExtensionCriticalScopes=$this.ControlSettings.Organization.ExtensionCriticalScopes;
+                        $ExtensionListWithCriticalScopes = $extensionList | Where-Object {$ExtensionCriticalScopes -contains $_.scopes }
+                        $controlResult.AddMessage("`nNo. of extensions have critical access permissions : "+ $ExtensionListWithCriticalScopes.count)                        
+                        $display= $ExtensionListWithCriticalScopes|Format-Table -Property  @{name="ExtensionName";expression={$_.extensionName}},@{name="Scope";expression={$_.scopes}} | Out-String
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:427]: List of extensions that have critical access permissions:

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 7]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                        $controlResult.AddMessage($display) 
+
+                        # Avoid extensions  with 'DevTest', 'Demo', 'Preview', 'Deprecated' in names
+                        $ExtensionListWithInvalidExtensionNames=@()
+                        $InvalidExtensionNames=$this.ControlSettings.Organization.InvalidExtensionNames;
+                        for($i=0;$i -lt $extensionList.count;$i++)
+                        {
+                            for($j=0;$j -lt $InvalidExtensionNames.Count;$j++)
+                            {
+                                if($extensionList[$i] -match $InvalidExtensionNames[$j])
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:437]: $extensionList[$i].name

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 8]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                                {
+                                    [System.Array] $ExtensionListWithInvalidExtensionNames = $extensionList[$i]
+                                }
+                            }
+                        }                        
+                        if($ExtensionListWithInvalidExtensionNames.count -gt 0)
+                        {
+                        $controlResult.AddMessage("`nExtensions having name as of test environment : "+ $ExtensionListWithInvalidExtensionNames.count)
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:445]: No. of extensions that have name belonging to non-production enevironment:

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 9]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                                    {
+                                        'criteria': [
+                                            {
+                                                'filterType': 7,
+                                                'value': '$($_.publisherId).$($_.extensionId)'
+                                            }
+                                        ]                                
+                                    }
+                                ]
+                            }" | ConvertFrom-Json | ConvertTo-json -Depth 10
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:470]: is the convulation from json to PSObject to json again reqd?

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 10]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                                -UseBasicParsing
+    
+                            $responseObject=$response.Content | ConvertFrom-Json
+
+                            if([Helpers]::CheckMember($responseobject.results[0], "extensions") -eq $false )
+                            {
+                                $privateExtensions+=$_
+                            }
+                            else{
+                                if($responseobject.results.extensions.flags -match "preview")
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:485]: add via control settings

Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:485]: results[0]

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 11]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                                    $nonProdExtensions+=$_
+                                }
+                                if($responseobject.results.extensions.publisher.flags -match "certified")
+                                {
+                                    $topPublisherExt+=$_
+                                }
+                            }                            
+                        }
+
+                        $controlResult.AddMessage("`nNo. of installed extensions with Top Publishers: "+$topPublisherExt.count);
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:496]: from top publishers:

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 12]
```diff
@@ -401,21 +401,119 @@ class Organization: ADOSVTBase
...
+                        $controlResult.AddMessage($topPublisherExt);
+                        $controlResult.AddMessage("`nNo. of installed extensions with private visibility: "+$privateExtensions.count);
+                        $controlResult.AddMessage($privateExtensions);
+                        $controlResult.AddMessage("`nNo. of installed extensions that are non Prod: "+$nonProdExtensions.count);
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:500]: that belong to non production environment

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 1 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: Added check for control settings

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 2 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: moved after basic detailed logs

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 3 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: updated message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 4 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: updated message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 5 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: Appended

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 6 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: Updated message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 7 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: Updated as per comments

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 8 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: Updated the message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 9 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: removed as convolution was not required

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 10 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: implemented via control settings

Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: added 0th index in results

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 11 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: updated the message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 12 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: updated the message

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 3 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: added if condition for count

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 13]
```diff
@@ -448,6 +454,134 @@ class Organization: ADOSVTBase
...
+    
+                                $response= Invoke-WebRequest -Uri $url `
+                                    -Method Post `
+                                    -ContentType "application/json" `
+                                    -Body $inputbody `
+                                    -UseBasicParsing
+    
+                                $responseObject=$response.Content | ConvertFrom-Json
+
+                                if([Helpers]::CheckMember($responseobject.results[0], "extensions") -eq $false )
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:538]: add comments

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 [hunk 14]
```diff
@@ -448,6 +454,134 @@ class Organization: ADOSVTBase
...
+                                    if($responseobject.results[0].extensions.publisher.flags -match "certified")
+                                    {
+                                        $topPublisherExt+=$_
+                                    }
+                                }                            
+                            }
+
+                            if($topPublisherExt.count -gt 0)
+                            {
+                                $controlResult.AddMessage("`nNo. of installed extensions from Top Publishers: "+$topPublisherExt.count);
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:566]: top publishers

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 13 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1:547]: Added comments

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1 (hunk 14 above)
Comment (v-himkam) [reply src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Organization.ps1]: lowered the cases

**Review by abhaydaga — APPROVED** [review]

**Pull request merged**